### PR TITLE
Add voice-over and voice control supports to the Dashboard page.

### LINKED ISSUE #51: IMPROVE - Voice Over / Voice Control Accessibility for `DashboardView`
## Describe the feature you want to improve
* Go through the `DashboardView` and check for proper Voice Over support and Voice Control

## Describe alternatives you've considered
* N/A

### PR DESCRIPTION
**Pull request by rbennum**
# What it Does
* Closes #51 
* Adds voice-over and voice control supports on the Dashboard page.

# How I Tested
* Run voice control and voice-over from the Accessibility menu on Settings.
* Open the app to the Dashboard page.
* Add and delete events on your own.

# Notes
* There is a bug I can't handle right now when I try to fill the bottom part of the text field in any views using Voice Control but the first responder is not activating.
* Also I'm still learning how to navigate swipeable cells using the Voice-Over feature so it just went untested for now.

# Screenshot

https://github.com/mikaelacaron/Basic-Car-Maintenance/assets/8705838/a4e5e8f6-d5ae-4ac6-8304-99ec9b019b0c

### CODE REVIEW

File: Basic-Car-Maintenance.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -14,6 +14,10 @@
 		57CDD9A02ADC31A8002EFED0 /* AddOdometerReadingView.swift in Sources */ = {isa = PBXBuildFile; fileRef = 57CDD99F2ADC31A8002EFED0 /* AddOdometerReadingView.swift */; };
 		57CDD9A42ADC320F002EFED0 /* OdometerReading.swift in Sources */ = {isa = PBXBuildFile; fileRef = 57CDD9A32ADC320F002EFED0 /* OdometerReading.swift */; };
 		637505482AEFFBAC00AA5D0B /* FirebaseAnalytics+Extension.swift in Sources */ = {isa = PBXBuildFile; fileRef = 637505472AEFFBAC00AA5D0B /* FirebaseAnalytics+Extension.swift */; };
+		6D7232312B42F7D400E0EEC8 /* Date+extension.swift in Sources */ = {isa = PBXBuildFile; fileRef = 6D7232302B42F7D400E0EEC8 /* Date+extension.swift */; };
```
Comment (rbennum) [inline Basic-Car-Maintenance.xcodeproj/project.pbxproj]: Have you considered using a build system like XcodeGen to help unclutter this kind of file?

Comment (mikaelacaron) [reply Basic-Car-Maintenance.xcodeproj/project.pbxproj]: Which thing are you referring to that is this cluttered? Why are there so many changes for one file?

File: Basic-Car-Maintenance/Shared/Utilities/Date+extension.swift [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+extension Date {
+    /**
+     A simple date-to-string converter with a predefined format style.
+     */
+    func toString(formatted: DateFormatter.Style = .short) -> String {
+        let formatter = DateFormatter()
+        formatter.dateStyle = .short
+        return formatter.string(from: self)
+    }
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Utilities/Date+extension.swift]: This function isn't using the input parameter `formatted` it's creating its own `formatter` every time you'd use it and right now it will always use `.short` as the format.

I think for now it's fine the way it was and we don't need this extension yet

File: Basic-Car-Maintenance/Shared/Dashboard/Views/AddMaintenanceView.swift [hunk 3]
```diff
@@ -22,54 +22,70 @@ struct AddMaintenanceView: View {
...
-                              prompt: Text("The title of the event",
-                                           comment: "Maintenance event title text field label placeholder")) {
+                    TextField(
+                        text: $title,
+                        prompt: Text(
+                            "The title of the event",
+                            comment: "Maintenance event title text field label placeholder"
+                        )
+                        .accessibilityLabel("Event title.")
+                    ) {
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Dashboard/Views/AddMaintenanceView.swift]: Don't put a period after `Event Title`

All the changes on this file are actually out of scope of your issue, because as noted in this [comment](https://github.com/mikaelacaron/Basic-Car-Maintenance/issues/51#issuecomment-1868991011) the changes you agreed to make were only going to be on the `DashboardView`

Can you remove all of the code that's outside the scope of the issue please

Also next time do **not** change how this was indented. Note previously it was indented this way
```swift
TextField(text: $title,
   prompt: Text("..."))
```

and you changed the way it was indented to:
```swift
TextField(
   text: $title,
   prompt: Text("...")
)
```

If you did this for every piece of code you touched, it would start to make the codebase inconsistent

File: Basic-Car-Maintenance/Shared/Dashboard/Views/EditEventDetailView.swift [hunk 4]
```diff
@@ -14,48 +14,70 @@ struct EditMaintenanceEventView: View {
     @State private var date = Date()
     @State private var notes = ""
     @Environment(\.dismiss) var dismiss
-    
+
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Dashboard/Views/EditEventDetailView.swift]: All the changes on this file are actually out of scope of your issue, because as noted in this [comment](https://github.com/mikaelacaron/Basic-Car-Maintenance/issues/51#issuecomment-1868991011) the changes you agreed to make were only going to be on the `DashboardView`

Can you remove all of the code that's outside the scope of the issue please

File: Basic-Car-Maintenance/Shared/Settings/Views/SettingsView.swift [hunk 5]
```diff
@@ -51,10 +51,12 @@ struct SettingsView: View {
                         }
                     }
                     .popoverTip(ContributionTip(), arrowEdge: .bottom)
+                    .accessibilityLabel("Link to the Basic Car Maintenance GitHub repo.")
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Settings/Views/SettingsView.swift]: All the changes on this file are actually out of scope of your issue, because as noted in this [comment](https://github.com/mikaelacaron/Basic-Car-Maintenance/issues/51#issuecomment-1868991011) the changes you agreed to make were only going to be on the `DashboardView`

Can you remove all of the code that's outside the scope of the issue please

File: Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift [hunk 6]
```diff
@@ -51,24 +50,29 @@ struct DashboardView: View {
...
                                 Image(systemName: SFSymbol.pencil)
                             }
                         }
                     }
                     .sheet(isPresented: $isShowingEditView) {
                         EditMaintenanceEventView(
-                            selectedEvent: $selectedMaintenanceEvent, viewModel: viewModel)
+                            selectedEvent: $selectedMaintenanceEvent, 
+                            viewModel: viewModel
+                        )
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift]: Don't change the indentation just because you see something, mention it in the PR that you think it should be changed. Or make an issue about several places where you think indentation should change

Because making "small changes" like this throughout a PR makes the PR bigger, harder to review, and loses focus on the actual issue you're solving

File: Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift [hunk 7]
```diff
@@ -51,24 +50,29 @@ struct DashboardView: View {
...
                     }
+                    .accessibilityElement(children: .combine)
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift]: what does adding this do? does it do something to the sheet? the whole swipe actions? or combines the whole `VStack`? which if it's the `VStack`, can you actually move this to be a modifier on the VStack, before the `swipeActions`

File: Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift [hunk 8]
```diff
@@ -81,6 +85,7 @@ struct DashboardView: View {
                     if viewModel.events.isEmpty {
                         Text("Add your first maintenance",
                              comment: "Placeholder text for empty maintenance list prompting the user to add a maintenance event") // swiftlint:disable:this line_length
+                        .accessibilityLabel("Add your first maintenance.")
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift]: Don't add a period, this makes it so you made a new entry in the Localizable.xcstrings file, when in reality it's not a new entry, so it creates more work for someone to localize this

File: Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift [hunk 9]
```diff
@@ -137,10 +152,12 @@ struct DashboardView: View {
                     .accessibilityShowsLargeContentViewer {
                         Label {
                             Text("AddEvent", comment: "Label for adding maintenance event on Dashboard view")
+                                .accessibilityLabel("Add an event.")
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift]: don't add a period

File: Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift [hunk 10]
```diff
@@ -181,21 +198,21 @@ struct DashboardView: View {
                 await viewModel.getMaintenanceEvents()
             }
         }
-        .alert(Text("An Error Occurred",
-                    comment: "Title for alert shown when adding maintenance event fails"),
-               isPresented: $viewModel.showAddErrorAlert) {
+        .alert(
+            Text("An Error Occurred", comment: "Title for alert shown when adding maintenance event fails")
+                .accessibilityLabel("An error occurred."),
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift]: don't add a period

File: Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift [hunk 11]
```diff
@@ -181,21 +198,21 @@ struct DashboardView: View {
...
+            isPresented: $viewModel.showAddErrorAlert
+        ) {
             Button(role: .cancel) {} label: {
                 Text("OK", comment: "Label to dismiss alert")
+                    .accessibilityLabel("OK")
             }
+            .accessibilityInputLabels(["Dismiss"])
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift]: There is `Dismiss` and `Dismiss Alert` choose one or the other

**Review by mikaelacaron — CHANGES_REQUESTED** [review]
I was only expecting you to make changes to the `DashboardView` not the 3 other views too, can you please remove those changes for now and I'll be making a new issue to add those back later, thanks!

I have several comments that need resolved. If you have any questions, you can comment directly on that thread, otherwise click the resolve button once you've addressed my comment. Once you fix everything, please click the re-review button

Re-review button

**Comment by rbennum** [comment]
Hi @mikaelacaron thank you so much for addressing the problem here. Sorry for being absent as I was not in the best way possible to handle a lot of things for the last two weeks. I'll try to address your concern soon.

File: Basic-Car-Maintenance/Shared/Dashboard/Views/AddMaintenanceView.swift (hunk 3 above)
Comment (rbennum) [reply Basic-Car-Maintenance/Shared/Dashboard/Views/AddMaintenanceView.swift]: Yes, but in my defense, there are two quite critical views tagging along with the `DashboardView` because without two of them, the `DashboardView` would not stand on its own. But perhaps I overstepped my bounds. I will take care of it.

File: Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift (hunk 6 above)
Comment (rbennum) [reply Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift]: I'll keep that in mind!

File: Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift (hunk 8 above)
Comment (rbennum) [reply Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift]: Noted.

File: Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift (hunk 7 above)
Comment (rbennum) [reply Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift]: This command right here is used to group all text contents of the children views so the voice-over feature would read the whole thing, not separate it by view. Also, you're right, moving it directly under the VStack makes it clearer that it belongs only to the VStack. I'm handling it right now. Thank you.

**Comment by rbennum** [comment]
After considering all your inputs, the only thing worth adding is the way I handled the voice-over/control on `VStack`. If there are any actions related to other views, perhaps it's best to make another PR for it. Looking forward to your next input!

**Review by mikaelacaron — APPROVED** [review]
> After considering all your inputs, the only thing worth adding is the way I handled the voice-over/control on `VStack`. If there are any actions related to other views, perhaps it's best to make another PR for it. Looking forward to your next input!

We could've kept your other changes related to the `DashboardView`, but removed the other stuff, but it's fine I'll re-write this issue in a different way to be done later and make it more clear about what I want next time!

Thank you for contributing to the app!
